File: web/src/api/dataset.ts
import { api, AUTH_INFO } from './config';

// 数据集接口相关类型定义
export interface CreateDatasetRequest {
  proj_id: number;
  dataset_name: string;
  data_type: number; // 1: 静态数据, 2: 时序数据
  freqs?: string; // 时序周期，如 "1m+5m+1H+1D"
  check_rules?: string; // 校验规则
  comment?: string; // 备注
}

export interface UpdateDatasetRequest {
  proj_id: number;
  dataset_id: number;
  dataset_name?: string;
  check_rules?: string;
  comment?: string;
}

export interface DeleteDatasetRequest {
  proj_id: number;
  dataset_id: number;
}

// 返回信息类型定义
export interface RetInfo {
  code: number;
  msg: string;
}

// 数据集响应类型定义
export interface DatasetResponse {
  ret_info: RetInfo;
  dataset_id?: number;
}

// 创建数据集
export const createDataset = async (params: CreateDatasetRequest): Promise<DatasetResponse> => {
  try {
    console.log('创建数据集请求参数:', params);
    
    const response = await api.post('/metadata/CreateDataSet', {
      auth_info: {
        app_id: AUTH_INFO.app_id,
        app_key: AUTH_INFO.app_key
      },
      ...params
    });

    console.log('创建数据集响应:', response.data);
    const data = response?.data;
    
    // 添加安全检查
    if (!data) {
      throw new Error('创建数据集失败：响应数据为空');
    }
    
    if (!data.ret_info) {
      throw new Error('创建数据集失败：响应格式错误，缺少ret_info字段');
    }
    
    // 检查ret_info.code是否为0（成功）
    if (data.ret_info.code !== 0) {
      throw new Error(data.ret_info.msg || '创建数据集失败');
    }
    
    return data;
  } catch (error: any) {
    console.error('创建数据集失败:', error);
    throw new Error(error?.message || '创建数据集失败');
  }
};

// 更新数据集
export const updateDataset = async (params: UpdateDatasetRequest): Promise<DatasetResponse> => {
  try {
    console.log('更新数据集请求参数:', params);
    
    const response = await api.post('/metadata/UpdateDataSet', {
      auth_info: {
        app_id: AUTH_INFO.app_id,
        app_key: AUTH_INFO.app_key
      },
      ...params
    });

    console.log('更新数据集响应:', response.data);
    const data = response?.data;
    
    // 添加安全检查
    if (!data) {
      throw new Error('更新数据集失败：响应数据为空');
    }
    
    if (!data.ret_info) {
      throw new Error('更新数据集失败：响应格式错误，缺少ret_info字段');
    }
    
    // 检查ret_info.code是否为0（成功）
    if (data.ret_info.code !== 0) {
      throw new Error(data.ret_info.msg || '更新数据集失败');
    }
    
    return data;
  } catch (error: any) {
    console.error('更新数据集失败:', error);
    throw new Error(error?.message || '更新数据集失败');
  }
};

// 删除数据集
export const deleteDataset = async (params: DeleteDatasetRequest): Promise<DatasetResponse> => {
  try {
    console.log('删除数据集请求参数:', params);
    
    const response = await api.post('/metadata/DeleteDataSet', {
      auth_info: {
        app_id: AUTH_INFO.app_id,
        app_key: AUTH_INFO.app_key
      },
      ...params
    });

    console.log('删除数据集响应:', response.data);
    const data = response?.data;
    
    // 添加安全检查
    if (!data) {
      throw new Error('删除数据集失败：响应数据为空');
    }
    
    if (!data.ret_info) {
      throw new Error('删除数据集失败：响应格式错误，缺少ret_info字段');
    }
    
    // 检查ret_info.code是否为0（成功）
    if (data.ret_info.code !== 0) {
      throw new Error(data.ret_info.msg || '删除数据集失败');
    }
    
    return data;
  } catch (error: any) {
    console.error('删除数据集失败:', error);
    throw new Error(error?.message || '删除数据集失败');
  }
}; 

File: modules/storage/internal/services/metadata/dao/sqlite/dataset.go
package sqlite

import (
	"fmt"

	"github.com/mooyang-code/moox/modules/storage/internal/services/common/constants"
	"github.com/mooyang-code/moox/modules/storage/internal/services/metadata/dao/model"
	"trpc.group/trpc-go/trpc-go/log"
)

// Dataset 数据集定义表
type Dataset model.Dataset

// TableName 指定表名
func (d Dataset) TableName() string {
	return model.DatasetTableName
}

// GetDatasetList 获取所有数据集列表
func (d *dataDBImpl) GetDatasetList() ([]model.Dataset, error) {
	var datasets []model.Dataset
	result := d.db.Where("c_enabled = ?", constants.EnabledValue).Find(&datasets)
	if result.Error != nil {
		log.Errorf("GetDatasetList err[%v]", result.Error)
		return nil, result.Error
	}
	return datasets, nil
}

// GetDatasetByID 根据ID获取数据集
func (d *dataDBImpl) GetDatasetByID(datasetID int) (*model.Dataset, error) {
	var dataset model.Dataset
	result := d.db.Where("c_dataset_id = ? AND c_enabled = ?", datasetID, constants.EnabledValue).First(&dataset)
	if result.Error != nil {
		log.Errorf("GetDatasetByID err[%v]", result.Error)
		return nil, result.Error
	}
	return &dataset, nil
}

// GetDatasetByProjID 根据项目ID获取数据集列表
func (d *dataDBImpl) GetDatasetByProjID(projID int) ([]model.Dataset, error) {
	var datasets []model.Dataset
	result := d.db.Where("c_proj_id = ? AND c_enabled = ?", projID, constants.EnabledValue).Find(&datasets)
	if result.Error != nil {
		log.Errorf("GetDatasetByProjID err[%v]", result.Error)
		return nil, result.Error
	}
	return datasets, nil
}

// GetDatasetByDataType 根据数据类型获取数据集列表
func (d *dataDBImpl) GetDatasetByDataType(dataType int) ([]model.Dataset, error) {
	var datasets []model.Dataset
	result := d.db.Where("c_data_type = ? AND c_enabled = ?", dataType, constants.EnabledValue).Find(&datasets)
	if result.Error != nil {
		log.Errorf("GetDatasetByDataType err[%v]", result.Error)
		return nil, result.Error
	}
	return datasets, nil
}

// AddDataset 添加新的数据集
func (d *dataDBImpl) AddDataset(dataset *model.Dataset) error {
	if dataset.Enabled == "" {
		dataset.Enabled = constants.EnabledValue
	}
	result := d.db.Create(dataset)
	if result.Error != nil {
		log.Errorf("AddDataset err[%v]", result.Error)
		return result.Error
	}
	return nil
}

// UpdateDataset 更新数据集信息
func (d *dataDBImpl) UpdateDataset(dataset *model.Dataset) error {
	result := d.db.Model(&model.Dataset{}).Where("c_dataset_id = ?", dataset.DatasetID).
		Updates(map[string]any{
			"c_dataset_name": dataset.DatasetName,
			"c_proj_id":      dataset.ProjID,
			"c_data_type":    dataset.DataType,
			"c_freqs":        dataset.Freqs,
			"c_check_rules":  dataset.CheckRules,
			"c_comment":      dataset.Comment,
		})
	if result.Error != nil {
		log.Errorf("UpdateDataset err[%v]", result.Error)
		return result.Error
	}
	return nil
}

// DeleteDataset 逻辑删除数据集
func (d *dataDBImpl) DeleteDataset(datasetID int) error {
	result := d.db.Model(&model.Dataset{}).Where("c_dataset_id = ?", datasetID).
		Update("c_enabled", constants.DisabledValue)
	if result.Error != nil {
		log.Errorf("DeleteDataset err[%v]", result.Error)
		return result.Error
	}
	return nil
}

// GetMaxDatasetIDInRange 获取指定范围内的最大数据集ID
func (d *dataDBImpl) GetMaxDatasetIDInRange(minID, maxID int) (int, error) {
	var result int
	err := d.db.Table("t_dataset").
		Where("c_dataset_id >= ? AND c_dataset_id <= ?", minID, maxID).
		Select("COALESCE(MAX(c_dataset_id), ?)", minID-1).
		Scan(&result).Error
	if err != nil {
		return 0, fmt.Errorf("获取最大数据集ID失败: %w", err)
	}
	return result, nil
}
